app/src/main/java/com/concordium/wallet/data/preferences/AuthPreferences.kt

```diff
@@ -142,17 +146,61 @@ class AuthPreferences(val context: Context) :
         setString(SEED_PHRASE, "")
     }
 
-    fun setSeedPhrase(value: String) {
+    suspend fun setSeedPhrase(value: String, password: String): Boolean {
         val seed = Mnemonics.MnemonicCode(value).toSeed()
         val seedEncoded = seed.toHex()
-        setString(SEED_PHRASE, seedEncoded)
-    }
-
-    fun getSeedPhrase(): String {
-        return getString(SEED_PHRASE, "")
+        val encryptedSeed = App.appCore.getCurrentAuthenticationManager().encryptInBackground(password, seedEncoded)
+            ?: return false
+       return setStringWithResult(SEED_PHRASE_ENCRYPTED, encryptedSeed)
+    }
+
+    suspend fun getSeedPhrase(password: String): String {
+        val seedEncrypted = getString(SEED_PHRASE_ENCRYPTED)
+            ?: getString(SEED_PHRASE, "").let {seedUnencrypted ->
+                if(seedUnencrypted != ""){
+                    setStringWithResult(SEED_PHRASE, null).let {deleteSuccess ->
+                        if(deleteSuccess){
+                            val encryptedSeed = App.appCore.getCurrentAuthenticationManager().encryptInBackground(password, seedUnencrypted)
+                                ?: return ""
+                            setStringWithResult(SEED_PHRASE_ENCRYPTED, encryptedSeed).let {saveSuccess ->
+                                if(saveSuccess){
+                                    Log.d("ENCRYPTED SEED UPDATED")
+                                    return  seedUnencrypted
+                                }
+                            }
+                        }
+                    }
+                }
+                    return ""
+            }
```

Comment:
I think it's important that the original seed phrase is deleted only after the encrypted one has been successfully stored.

I also feel like this would be easier to read without such heavy nesting. I'm not aware of common Kotlin style, but I would find something like this much clearer

```
        val seedEncrypted = getString(SEED_PHRASE_ENCRYPTED)
        if (seedEncrypted != null) {
            // happy path
            return App.appCore.getCurrentAuthenticationManager().decryptInBackground(password, seedEncrypted) ?: ""
        }
        val seedUnencrypted = getString(SEED_PHRASE, "")
        if (seedUnencrypted == "") {
            return ""
        }
        val encryptedSeed = App.appCore.getCurrentAuthenticationManager().encryptInBackground(password, seedUnencrypted)
        if (encryptedSeed == null) {
            return ""
        }
        val saveSuccess = setStringWithResult(SEED_PHRASE_ENCRYPTED, encryptedSeed)
        if (saveSuccess) {
            Log.d("ENCRYPTED SEED UPDATED")
        }
        val deleteSuccess = setStringWithResult(SEED_PHRASE, null)
        if (deleteSuccess) {
            Log.d("UNENCRYPTED SEED DELETED")
        }
        return seedUnencrypted
```

Finally, what happens in all the cases where the empty string is returned? Aren't they all cases where we should actually report an error somehow? Same with the `else` cases where store/delete fail.

Reply:
I would even consider adding some further insurance that the seed phrase isn't lost if something goes wrong.

The most forcing (but also visible) way would be to ensure that user has the seed phrase offline either by asking them to confirm or doing an actual recovery flow.

A less user invasive option is to not delete the unencrypted value immediately, but do it on the happy path in the next invocation. That is, once the encrypted phrase has been successfully fetched and decrypted.

Reply:
The empty string triggers an error that is shown to the user later in the App. This is how it was previously set-up

I considered to keep the original seed phrase until the encrypted one is saved, but then if the encryption or saving fails the seed would remain unencrypted on the device until the next invocation. 

Maybe we can do this when the user is asked for the passcode when starting the app, and if the process fails we can show him an error and prevent him from entering the app and asking him to enter the passcode again.

Reply:
> but then if the encryption or saving fails the seed would remain unencrypted on the device until the next invocation.

That's what I would expect to happen. Is it better that there is no seed phrase than only the unencrypted one?